dds/DCPS/transport/framework/ReceivedDataSample.cpp

```diff
@@ -6,9 +6,169 @@
  */
 
 #include "DCPS/DdsDcps_pch.h" //Only the _pch include should start with DCPS/
+
 #include "ReceivedDataSample.h"
-#include "EntryExit.h"
 
-#if !defined (__ACE_INLINE__)
-# include "ReceivedDataSample.inl"
-#endif /* !__ACE_INLINE__ */
+#include <algorithm>
+
+#include <cstring>
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+namespace OpenDDS {
+namespace DCPS {
+
+ReceivedDataSample::ReceivedDataSample(const ACE_Message_Block& payload)
+{
+  const ACE_Message_Block* amb = &payload;
+  do {
+    blocks_.push_back(MessageBlock(*amb));
+    amb = amb->cont();
+  } while (amb);
+}
+
+size_t ReceivedDataSample::data_length() const
```

Comment:
It's certainly possible to cache it since the data is really encapsulated within private members of this class.

Of course it's hard to say what the performance implications will be, but unlike ACE_Message_Block itself there's no linked list pointer chasing which is cache-unfriendly.  All of these wr_ptr and rd_ptr values are within the contiguous allocation of our vector.

At least for now, `data_length()` is not called on the critical path.